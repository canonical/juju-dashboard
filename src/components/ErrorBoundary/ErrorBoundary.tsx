import {
  CodeSnippet,
  CodeSnippetBlockAppearance,
  Notification,
  Strip,
} from "@canonical/react-components";
import * as Sentry from "@sentry/browser";
import type { Extras } from "@sentry/core";
import type { PropsWithChildren } from "react";
import { Component } from "react";

import { externalURLs } from "urls";

type Props = PropsWithChildren;

type State = {
  error?: Error;
  eventId?: string;
  hasError: boolean;
};

export default class ErrorBoundary extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError(error: Error) {
    return {
      hasError: true,
      error,
    };
  }

  componentDidCatch(error: Error, info: unknown) {
    if (import.meta.env.PROD && window.jujuDashboardConfig?.analyticsEnabled) {
      Sentry.withScope((scope) => {
        scope.setExtras(info as Extras);
        const eventId = Sentry.captureException(error);
        this.setState({ eventId });
      });
    }
  }

  render() {
    const { error, hasError } = this.state;
    const { children } = this.props;
    const body = encodeURIComponent(
      `\`\`\`\n${error?.stack ?? "No stack track"}\n\`\`\``,
    );
    const url = `${
      externalURLs.newIssue
    }?assignees=&labels=&template=bug_report.md&title=Dashboard error: ${encodeURIComponent(
      error?.message ?? "No error",
    )}&body=${body}`;
    if (hasError) {
      return (
        <Strip>
          <Notification severity="negative" title="Error">
            Something has gone wrong. If this issue persists,{" "}
            <a href={url} rel="noopener noreferrer" target="_blank">
              please raise an issue on GitHub.
            </a>
          </Notification>
          <CodeSnippet
            blocks={[
              ...(error?.message
                ? [
                    {
                      title: "Error",
                      appearance: CodeSnippetBlockAppearance.NUMBERED,
                      wrapLines: true,
                      code: error.message,
                    },
                  ]
                : []),
              ...(error?.stack
                ? [
                    {
                      title: "Stack trace",
                      appearance: CodeSnippetBlockAppearance.NUMBERED,
                      wrapLines: true,
                      code: error.stack,
                    },
                  ]
                : []),
            ]}
          />
          <p>
            Further debugging can be done{" "}
            <a
              href={externalURLs.troubleshootDeployment}
              rel="noopener noreferrer"
              target="_blank"
            >
              via the Juju CLI
            </a>
            .
          </p>
        </Strip>
      );
    }

    return <>{children}</>;
  }
}
