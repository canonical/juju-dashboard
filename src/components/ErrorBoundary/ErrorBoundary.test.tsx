import { render } from "@testing-library/react";
import type { ReactNode } from "react";
import type { MockInstance } from "vitest";
import { vi } from "vitest";

import ErrorBoundary from "./ErrorBoundary";

function ChildComponent(): ReactNode {
  const error = new Error("Oh noes!");
  error.stack = "Stack trace output";
  throw error;
  return null;
}

describe("Error Boundary", () => {
  let consoleError: MockInstance;

  beforeEach(() => {
    // Vitest does not catch the error thrown in the component so this cleans up
    // the console output.
    consoleError = vi.spyOn(console, "error").mockImplementation(() => vi.fn());
  });

  afterEach(() => {
    consoleError.mockRestore();
  });

  it("renders without crashing and matches snapshot", () => {
    const { container } = render(
      <ErrorBoundary>
        <ChildComponent />
      </ErrorBoundary>,
    );
    expect(container).toMatchSnapshot();
  });
});
