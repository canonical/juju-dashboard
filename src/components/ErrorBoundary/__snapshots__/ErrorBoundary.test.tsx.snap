// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Error Boundary > renders without crashing and matches snapshot 1`] = `
<div>
  <div
    class="p-strip"
  >
    <div
      class="row"
    >
      <div
        class="col-12"
      >
        <div
          class="p-notification--negative"
        >
          <div
            class="p-notification__content"
          >
            <h5
              class="p-notification__title"
              data-testid="notification-title"
            >
              Error
            </h5>
            <p
              class="p-notification__message"
            >
              Something has gone wrong. If this issue persists,
               
              <a
                href="https://github.com/canonical/juju-dashboard/issues/new?assignees=&labels=&template=bug_report.md&title=Dashboard error: Oh%20noes!&body=%60%60%60%0AStack%20trace%20output%0A%60%60%60"
                rel="noopener noreferrer"
                target="_blank"
              >
                please raise an issue on GitHub.
              </a>
            </p>
          </div>
        </div>
        <div
          class="p-code-snippet"
        >
          <div
            class="p-code-snippet__header "
          >
            <h5
              class="p-code-snippet__title"
            >
              Error
            </h5>
          </div>
          <pre
            class="p-code-snippet__block--numbered is-wrapped"
          >
            <code>
              <span
                class="p-code-snippet__line"
              >
                Oh noes!
              </span>
              

            </code>
          </pre>
          <div
            class="p-code-snippet__header "
          >
            <h5
              class="p-code-snippet__title"
            >
              Stack trace
            </h5>
          </div>
          <pre
            class="p-code-snippet__block--numbered is-wrapped"
          >
            <code>
              <span
                class="p-code-snippet__line"
              >
                Stack trace output
              </span>
              

            </code>
          </pre>
        </div>
        <p>
          Further debugging can be done
           
          <a
            href="https://documentation.ubuntu.com/juju/latest/howto/manage-your-deployment/index.html#troubleshoot-your-deployment"
            rel="noopener noreferrer"
            target="_blank"
          >
            via the Juju CLI
          </a>
          .
        </p>
      </div>
    </div>
  </div>
</div>
`;
